node/service/src/lib.rs

```diff
@@ -899,16 +906,56 @@ where
 		slot_duration_millis: slot_duration.as_millis() as u64,
 	};
 
+	let (prep_worker_path, exec_worker_path) = if let Some(path) = workers_path {
+		log::trace!("Using explicitly provided workers path {:?}", path);
+		if path.is_executable() {
+			(path.clone(), path)
+		} else {
+			let mut prep_worker = path.clone();
+			let mut exec_worker = path.clone();
+			prep_worker.push(polkadot_node_core_pvf::PREPARE_BINARY_NAME);
+			exec_worker.push(polkadot_node_core_pvf::EXECUTE_BINARY_NAME);
+			(prep_worker, exec_worker)
+		}
+	} else {
+		let libexec = PathBuf::from("/usr/libexec");
+		let mut prep_worker = libexec.clone();
+		prep_worker.push(polkadot_node_core_pvf::PREPARE_BINARY_NAME);
+		let mut exec_worker = libexec.clone();
+		exec_worker.push(polkadot_node_core_pvf::EXECUTE_BINARY_NAME);
+
+		if prep_worker.exists() && exec_worker.exists() {
+			log::trace!("Using /usr/libexec as workers path");
+			(prep_worker, exec_worker)
+		} else {
+			let mut exe_path = std::env::current_exe()?;
+			let _ = exe_path.pop();
+			let mut prep_worker = exe_path.clone();
+			prep_worker.push(polkadot_node_core_pvf::PREPARE_BINARY_NAME);
+			let mut exec_worker = exe_path.clone();
+			exec_worker.push(polkadot_node_core_pvf::EXECUTE_BINARY_NAME);
+
+			if prep_worker.exists() && exec_worker.exists() {
+				log::trace!("Using current exe path as workers path: {:?}", exe_path);
+				(prep_worker, exec_worker)
+			} else {
+				log::trace!("Workers path not found, considering `$PATH`");
+				(
+					PathBuf::from(polkadot_node_core_pvf::EXECUTE_BINARY_NAME),
+					PathBuf::from(polkadot_node_core_pvf::PREPARE_BINARY_NAME),
```

Comment:
Maybe we should check if these exist and are executable and not assume so. If something is wrong, emit an error and bail? We shouldn't continue without valid workers. Also, would be good to bail on borked installations as people may run into weird issues otherwise.

Speaking of which, maybe we should do the version mismatch check right after this, i.e. run the workers with `--version` argument and make sure it's what we expect.

So I'd change this code to:

1. Go in order and get first set of paths which exist.
2. If no paths exist, error out.
3. Check if the returned paths are executable, if not it's evidence of a borked installation so error out.
4. Do the version check, if mismatch error out.
5. At this point the paths should be good to use.

Reply:
@s0me0ne-unkn0wn I'm thinking of removing the usage of `$PATH` as a last resort:

1. `is_executable` doesn't work without the full filepath, i.e. if just `polkadot-execute-worker` is passed, so it would complicate the code to add the check for (2) and (3) above.
2. I think we should make it standard for users to place the binaries in the same directory as the main polkadot executable. Otherwise, say I try to install manually and put new binaries in $PATH, but there are already old binaries in polkadot's directory -- I'd have a borked install. We could check this case programmatically but it would be more complication.

Seems simpler both for us and for users to just rely on predictable file locations, WDYT? Is $PATH a requirement for any use cases?

Reply:
> If not specified, the main binary's directory is searched first, then `/usr/libexec` is searched

I think we should always search both locations, and if the binaries are found at both it could indicate a bad install.

Also I think we should check for partial installs, e.g. execute worker exists but prepare worker doesn't. It's a minor complication, but we've spent a lot of time considering the user UX, so I want to try to get it right.

Reply:
> The version checking is excessive to my mind as we have to check versions on every worker execution anyway (it may be overwritten)

Yeah but I think it would be good UX to fail as early as possible.

> The question is if we want to force him to clean up after himself. I mean it's not our business how he manages his filesystem. 

Maybe it's enough to log a warning: `two sets of worker binaries found (<location X> and <location Y>), using the one at <location X>`.